src/envs/pddl_env.py

```diff
@@ -0,0 +1,492 @@
+"""Environments that are derived from PDDL.
+
+There are no continuous aspects of the state or action space. These
+environments are similar to PDDLGym.
+"""
+from __future__ import annotations
+
+import abc
+import functools
+from typing import Callable, Collection, Dict, List, Optional, Sequence, Set, \
+    Tuple, cast
+
+import numpy as np
+from gym.spaces import Box
+from pyperplan.pddl.parser import TraversePDDLDomain, TraversePDDLProblem, \
+    parse_domain_def, parse_lisp_iterator, parse_problem_def
+from pyperplan.pddl.pddl import Domain as PyperplanDomain
+
+from predicators.src import utils
+from predicators.src.envs import BaseEnv
+from predicators.src.envs.pddl_procedural_generation import \
+    create_blocks_pddl_generator
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, GroundAtom, Image, \
+    LiftedAtom, Object, ParameterizedOption, PDDLProblemGenerator, Predicate, \
+    State, STRIPSOperator, Task, Type, Variable, _GroundSTRIPSOperator
+
+###############################################################################
+#                                Base Classes                                 #
+###############################################################################
+
+
+class _PDDLEnvState(State):
+    """No continuous object features, and ground atoms in simulator_state."""
+
+    def get_ground_atoms(self) -> Set[GroundAtom]:
+        """Expose the ground atoms in the simulator_state."""
+        return cast(Set[GroundAtom], self.simulator_state)
+
+    @classmethod
+    def from_ground_atoms(cls, ground_atoms: Set[GroundAtom],
+                          objects: Collection[Object]) -> _PDDLEnvState:
+        """Create a state from ground atoms and objects."""
+        # Keep a dummy state dict so we know what objects are in the state.
+        dummy_state_dict = {o: np.zeros(0, dtype=np.float32) for o in objects}
+        return _PDDLEnvState(dummy_state_dict, simulator_state=ground_atoms)
+
+    def allclose(self, other: State) -> bool:
+        return self.simulator_state == other.simulator_state
+
+    def copy(self) -> State:
```

Comment:
so we're not using the base copy anymore?

Reply:
that's right